llvm/test/CodeGen/RISCV/rvZce/cm_mvas_mvsa.ll

```diff
@@ -0,0 +1,163 @@
+
+define i64 @cmvs32(i64 %a) nounwind {
+;  VALID32-LABEL: cmvs32:
+;  VALID32: cm.mvsa01 {{s[0-7]}}, {{s[0-7]}}
+;  VALID32-NOT: cm.mvsa01 {{a.}}, {{a.}}
+  %tmp = call i64 @llvm.cttz.i64(i64 %a, i1 false)
+  ret i64 %tmp
+}
+
+declare void @hoge()
+define void @cmvs64(i32 signext %arg, i32 signext %arg1) nounwind {
+;  VALID64-LABEL: cmvs:
+;  VALID64: cm.mvsa01 {{s[0-7]}}, {{s[0-7]}}
+;  VALID64-NOT: cm.mvsa01 {{a.}}, {{a.}}
+bb:
+  %tmp = icmp eq i32 %arg, %arg1
+  br i1 %tmp, label %bb6, label %bb2
+
+bb2:                                              ; preds = %bb2, %bb
+  %tmp3 = phi i32 [ %tmp4, %bb2 ], [ %arg, %bb ]
+  tail call void @hoge()
+  %tmp4 = add nsw i32 %tmp3, 1
+  %tmp5 = icmp eq i32 %tmp4, %arg1
+  br i1 %tmp5, label %bb6, label %bb2
+
+bb6:                                              ; preds = %bb2, %bb
+  ret void
+}
+
+%struct.trie = type { [26 x %struct.trie*], i8 }
+
+@word = external global i8*
+
+declare i32 @trie_new(%struct.trie*)
+declare i32 @trie_search(i8*, i32, %struct.trie**)
+declare i64 @strnlen(i8*, i64)
+
+; Function Attrs: nounwind optnone
+define i32 @mvas_2() {
+  ;  VALID64-LABEL: mvas_2:
+  ;  VALID64-NOT: cm.mva01s {{a.}}, {{s.}}
+  ;  VALID64-NOT: cm.mva01s {{s.}}, {{a.}}
+entry:
+  %trie = alloca %struct.trie*
+  %0 = bitcast %struct.trie** %trie to i8*
+  store %struct.trie* null, %struct.trie** %trie
+  %call = tail call i32 @trie_new(%struct.trie* null)
+  %1 = load i8*, i8** @word
+  %call1 = tail call i64 @strnlen(i8* %1, i64 100)
+  %conv = trunc i64 %call1 to i32
+  %call2 = call i32 @trie_search(i8* %1, i32 %conv, %struct.trie** %trie)
+  ret i32 %call2
+}
+
+declare i32 @foo(i32, i32)
+
+; Function Attrs: nounwind optnone
+define dso_local i32 @cm_mvas_same_src(i32 %0, i32 %1, i32 %2, i32 %3) {
+ ;  VALID32-LABEL: cm_mvas_same_src:
+ ;  VALID32: cm.mva01s	s0, s0
+ ;
+ ;  VALID64-LABEL: cm_mvas_same_src:
+ ;  VALID64: cm.mva01s	s0, s0
+entry:
+  %4 = call i32 @foo(i32 %3, i32 %2)
+  %5 = add i32 %4, %2
+  %6 = call i32 @foo(i32 %3, i32 %3)
+  %add = add i32 %5, %6
+  ret i32 %add
+}
+
+%struct._IO_FILE = type { i32, i8*, i8*, i8*, i8*, i8*, i8*, i8*, i8*, i8*, i8*, i8*, %struct._IO_marker*, %struct._IO_FILE*, i32, i32, i64, i16, i8, [1 x i8], i8*, i64, i8*, i8*, i8*, i8*, i64, i32, [20 x i8] }
+%struct._IO_marker = type { %struct._IO_marker*, %struct._IO_FILE*, i32 }
+%struct.Node = type { i8*, i64, %struct.Node*, %struct.Node* }
+
+declare i8* @malloc(i64)
+
+declare i32 @fgetc(%struct._IO_FILE*nocapture)
+
+declare %struct.Node* @addWordToTree(i8*, %struct.Node*)
+
+; Function Attrs: nounwind optsize
+define %struct.Node* @cmmv_a1_come_first(%struct._IO_FILE*nocapture %file, %struct.Node* %root) {
+ ;  VALID32-LABEL: cmmv_a1_come_first:
+ ;  VALID32: cm.push	{ra, s0-s5}, -32
+ ;  VALID32-NEXT: .cfi_offset ra, -4
+ ;  VALID32-NEXT: .cfi_offset s0, -8
+ ;  VALID32-NEXT: .cfi_offset s1, -12
+ ;  VALID32-NEXT: .cfi_offset s2, -16
+ ;  VALID32-NEXT: .cfi_offset s3, -20
```

Comment:
Is .cfi_offset an assembly instruction? It's not recognized by llvm-mc?

echo ".cfi_offset ra, -4" |./llvm-mc --triple=riscv32 -mattr=experimental-zcmp 
        .text
<stdin>:1:1: error: this directive must appear between .cfi_startproc and .cfi_endproc directives
.cfi_offset ra, -4
^
        .cfi_offset ra, -4

Reply:
Hi @ChunyuLiao ,cfi_offset should be in the middle of .cfi_startproc and .cfi_endproc. They contain call frame infornation for debugging call frame stuff. I guess there is an attribute that can stop emiting cfi info, and I can have a look tomorrow.

Reply:
I checked all the assembler's file , .cfi_offset is between .cfi_startproc and .cfi_endproc.

Thanks for your contribution